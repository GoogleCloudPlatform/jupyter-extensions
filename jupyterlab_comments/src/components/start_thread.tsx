/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { useState } from 'react';
import { TextField, Grid, Input, Button } from '@material-ui/core';
import {
  newDetachedCommentThread,
  newReviewCommentThread,
} from '../service/add_comment';
import { SendButton } from './send_button';

interface Props {
  serverRoot: string;
  currFilePath: string;
  commentType: 'detached' | 'review';
  reviewHash?: string;
}

const style = {
  editor: {
    padding: 20,
  },
  submit: {
    paddingLeft: 10,
    display: 'inlineBlock',
  },
  textField: {
    width: 400,
  },
  submitOptions: {
    paddingLeft: 250,
  },
};

export function CommentEditor(props) {
  const [comment, setComment] = useState('');
  const [lineNumber, setLineNumber] = useState(0);
  const [showLineInput, setShowLineInput] = useState(false);

  const handleSubmit = event => {
    event.preventDefault();
    switch (props.commentType) {
      case 'review':
        newReviewCommentThread(
          props.currFilePath,
          props.serverRoot,
          comment,
          props.reviewHash,
          lineNumber
        );
        break;
      case 'detached':
        newDetachedCommentThread(
          props.currFilePath,
          props.serverRoot,
          comment,
          lineNumber
        );
        break;
    }
    setComment(''); //clear comment editor field
    setLineNumber(0);
    setShowLineInput(false);
  };
  return (
    <form
      onSubmit={handleSubmit}
      style={style.editor}
      className="commentSubmit"
    >
      <link
        rel="stylesheet"
        href="https://fonts.googleapis.com/icon?family=Material+Icons"
      />
      <Grid container direction="column">
        <Grid item>
          <TextField
            multiline
            rows={3}
            label="Start a new comment thread"
            value={comment}
            onChange={e => setComment(e.target.value)}
            variant="outlined"
            size="medium"
            style={style.textField}
            className="newThreadTextField"
          />
        </Grid>
        <Grid container direction="row" style={style.submitOptions}>
          {showLineInput ? (
            <Input
              type="number"
              value={lineNumber}
              onChange={e => setLineNumber(parseInt(e.target.value))}
              name="line number"
              margin="none"
              inputProps={{
                style: {
                  width: '40px',
                },
                min: 0,
              }}
            />
          ) : (
            <Button size="small" onClick={() => setShowLineInput(true)}>
              {'Set line #'}
            </Button>
          )}
          <Grid item style={style.submit}>
            <SendButton type="sendThread" />
          </Grid>
        </Grid>
      </Grid>
    </form>
  );
}

export class NewCommentThread extends React.Component<Props> {
  render() {
    return <CommentEditor {...this.props} />;
  }
}
