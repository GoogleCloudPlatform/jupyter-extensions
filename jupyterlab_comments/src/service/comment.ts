/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { timeAgo } from './timestamp';

export type DetachedComment = Comment;

export interface CodeReviewComment extends Comment {
  request: ReviewRequest;
  revision: any;
}

export interface Comment {
  author: any;
  text: any;
  timestamp: string;
  range: any;
  hash: any;
  filePath: string;
  children?: any;
  parent?: any;
}

export interface ReviewRequest {
  timestamp: any;
  reviewRef: string;
  targetRef: string;
  requester: string;
  description: any;
  baseCommit: any;
  reviewHash: any;
}

export function createDetachedCommentFromJSON(
  obj: any,
  filePath: string
): DetachedComment {
  const content = obj.comment;
  const hash = obj.hash;
  const children = obj.children;
  const now = new Date();
  const current = now.getTime();
  const timestampString = timeAgo(current, parseInt(content.timestamp) * 1000);
  const comment: DetachedComment = {
    author: content.author,
    text: content.description,
    timestamp: timestampString,
    range: content.location.range,
    hash: hash,
    filePath: filePath,
  };
  if (children) {
    comment.children = children;
  }
  if (content.parent) {
    comment.parent = parent;
  }
  return comment;
}

export function createReviewCommentFromJSON(
  obj: any,
  revision: any,
  request: any,
  filePath: string
): CodeReviewComment {
  const content = obj.comment;
  const hash = obj.hash;
  const children = obj.children;
  const now = new Date();
  const current = now.getTime();
  const timestampString = timeAgo(current, parseInt(content.timestamp) * 1000);
  request.reviewHash = revision;
  const comment: CodeReviewComment = {
    author: content.author,
    text: content.description,
    timestamp: timestampString,
    range: content.location.range,
    hash: hash,
    revision: revision,
    request: request,
    filePath: filePath,
  };
  if (children) {
    comment.children = children;
  }
  if (content.parent) {
    comment.parent = parent;
  }
  return comment;
}
