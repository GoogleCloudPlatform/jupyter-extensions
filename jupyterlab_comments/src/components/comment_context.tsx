/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/
import React from 'react';
import { RegularFile, NotebookFile, Range } from '../service/file';

interface Props {
  range: Range;
  filePath: string;
  file?: RegularFile | NotebookFile;
}

const styles = {
  lines: {
    display: 'block',
    backgroundColor: 'lightgrey',
    padding: '10px',
    borderRadius: '3px',
  },
  lineNumber: {
    color: 'white',
    display: 'inline',
    paddingRight: '5px',
    fontSize: '10px',
  },
  lineText: {
    display: 'inline',
  },
};

export class CommentContext extends React.Component<Props> {
  render() {
    const context = this.props.file.getContext(this.props.range);
    if (context) {
      const lines = context.map((line, index) => (
        <div key={index}>
          <p style={styles.lineNumber}> {line[0]} </p>
          <code style={styles.lineText}> {line[1]} </code>
        </div>
      ));
      return <div style={styles.lines}> {lines} </div>;
    } else {
      return <div> </div>;
    }
  }
}
