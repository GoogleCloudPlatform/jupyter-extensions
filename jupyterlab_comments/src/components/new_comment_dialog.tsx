/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { useState } from 'react';
import {
  Button,
  TextField,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
} from '@material-ui/core';
import { ReactWidget } from '@jupyterlab/apputils';
import { getServerRoot, Context } from '../service/jupyterConfig';
import { newDetachedCommentThread } from '../service/add_comment';

const style = {
  textField: {
    width: 400,
  },
};

function NewCommentDialog(props) {
  const [open, setOpen] = useState(true);
  const [comment, setComment] = useState('');

  const handleCloseCancel = () => {
    setOpen(false);
  };

  const handleCloseSend = () => {
    setOpen(false);
    const lineNumber = props.selection.start.line;
    newDetachedCommentThread(
      props.currFilePath,
      props.serverRoot,
      comment,
      lineNumber
    );
  };

  return (
    <div>
      <Dialog
        open={open}
        onClose={handleCloseCancel}
        aria-labelledby="form-dialog-title"
      >
        <DialogTitle id="form-dialog-title">Add a comment</DialogTitle>
        <DialogContent>
          <TextField
            autoFocus
            multiline
            margin="dense"
            rows={3}
            value={comment}
            variant="outlined"
            size="medium"
            label="Start a new comment thread"
            onChange={e => setComment(e.target.value)}
            style={style.textField}
            fullWidth
          />
        </DialogContent>
        <DialogActions>
          <Button onClick={handleCloseCancel} color="primary">
            Cancel
          </Button>
          <Button onClick={handleCloseSend} color="primary">
            Send
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  );
}

interface Props {
  serverRoot: string;
  currFilePath: string;
  commentType: 'detached' | 'review';
  selection: object;
}

class DialogComponent extends React.Component<Props> {
  render() {
    return <NewCommentDialog {...this.props} />;
  }
}

export class NewCommentDialogWidget extends ReactWidget {
  context: Context;
  selection: object;

  constructor(selection: object, context: Context) {
    super();
    this.selection = selection;
    this.context = context;
  }
  render() {
    const serverRoot = getServerRoot();
    const currWidget = this.context.labShell.currentWidget;
    const filePath = this.context.docManager.contextForWidget(currWidget).path;
    return (
      <DialogComponent
        selection={this.selection}
        serverRoot={serverRoot}
        commentType="detached"
        currFilePath={filePath}
      />
    );
  }
}
