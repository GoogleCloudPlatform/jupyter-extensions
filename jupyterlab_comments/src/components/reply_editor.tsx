/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/
import React, { useState } from 'react';
import { TextField, Grid } from '@material-ui/core';
import {
  newDetachedCommentReply,
  newReviewCommentReply,
} from '../service/add_comment';
import { getServerRoot } from '../service/jupyterConfig';
import { SendButton } from './send_button';

interface Props {
  currFilePath: string;
  hash: string;
  commentType: 'detached' | 'review';
  reviewHash?: string;
}

const style = {
  editor: {
    padding: 10,
  },
  textField: {
    margin: 2,
    width: 400,
  },
  submit: {
    paddingLeft: 320,
    display: 'inlineBlock',
  },
};

function ReplyEditor(props) {
  const [comment, setComment] = useState('');
  const serverRoot = getServerRoot();

  const handleSubmit = event => {
    event.preventDefault();
    switch (props.commentType) {
      case 'review':
        newReviewCommentReply(
          props.currFilePath,
          serverRoot,
          comment,
          props.hash,
          props.reviewHash
        );
        break;
      case 'detached':
        newDetachedCommentReply(
          props.currFilePath,
          serverRoot,
          comment,
          props.hash
        );
        break;
    }
    setComment(''); //clear comment text field
  };
  return (
    <form
      onSubmit={handleSubmit}
      style={style.editor}
      className="commentSubmit"
    >
      <link
        rel="stylesheet"
        href="https://fonts.googleapis.com/icon?family=Material+Icons"
      />
      <Grid container direction="column">
        <Grid item>
          <TextField
            multiline
            label="Reply to this comment"
            value={comment}
            rows={2}
            onChange={e => setComment(e.target.value)}
            variant="outlined"
            size="medium"
            style={style.textField}
            className="replyCommentTextField"
          />
        </Grid>
        <Grid item>
          <SendButton type="sendReply" style={style.submit} />
        </Grid>
      </Grid>
    </form>
  );
}

export class NewReplyComment extends React.Component<Props> {
  render() {
    return <ReplyEditor {...this.props} />;
  }
}
