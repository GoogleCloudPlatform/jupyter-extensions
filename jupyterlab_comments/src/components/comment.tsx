/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import {
  DetachedComment,
  createDetachedCommentFromJSON,
  CodeReviewComment,
  createReviewCommentFromJSON,
} from '../service/comment';
import {
  ListItem,
  ListItemText,
  ListItemAvatar,
  Avatar,
  Typography,
  Button,
  List,
} from '@material-ui/core';

import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';
import ArrowDropUpIcon from '@material-ui/icons/ArrowDropUp';
import { NewReplyComment } from './reply_editor';
import { CommentContext } from './comment_context';
import { RegularFile, NotebookFile } from '../service/file';

const style = {
  inline: {
    display: 'inline',
  },
  username: {
    fontWeight: 500,
    fontSize: 15,
    display: 'inline',
    textAlign: 'left' as const,
  },
  date: {
    display: 'inline',
    textAlign: 'left' as const,
    paddingLeft: 10,
    fontSize: 10,
    color: 'grey',
  },
  threadIndent: {
    paddingLeft: 50,
  },
  commentBottom: {
    display: 'inlineBlock',
    paddingLeft: 60,
  },
  replyButton: {
    paddingLeft: 60,
  },
};

interface Props {
  detachedComment?: DetachedComment;
  reviewComment?: CodeReviewComment;
  file?: RegularFile | NotebookFile;
}

interface State {
  expandThread: boolean;
  showCommentEditor: boolean;
  commentType: 'detached' | 'review';
}

//React component to render a single comment thread
export class Comment extends React.Component<Props, State> {
  constructor(props) {
    super(props);
    const commentType = this.props.detachedComment ? 'detached' : 'review';
    this.state = {
      expandThread: false,
      showCommentEditor: false,
      commentType: commentType,
    };
  }

  render() {
    const data = this.props.detachedComment
      ? this.props.detachedComment
      : this.props.reviewComment;

    return (
      <>
        <ListItem key="context">
          <CommentContext
            range={data.range}
            filePath={data.filePath}
            file={this.props.file}
          />
        </ListItem>
        <ListItem key="comment" alignItems="flex-start">
          <ListItemAvatar>
            <Avatar alt="avatar" />
          </ListItemAvatar>
          <ListItemText
            primary={
              <div>
                <p style={style.username}> {data.author} </p>
                <p style={style.date}> {data.timestamp} </p>
              </div>
            }
            secondary={
              <Typography
                variant="body2"
                style={style.inline}
                color="textPrimary"
              >
                {data.text}
              </Typography>
            }
          />
        </ListItem>
        <div style={style.commentBottom}>
          <Button
            color="primary"
            size="small"
            onClick={() => {
              this.setState({
                showCommentEditor: !this.state.showCommentEditor,
              });
            }}
            className="replyButton"
          >
            {' '}
            Reply{' '}
          </Button>
          {this.state.showCommentEditor &&
            (this.state.commentType === 'detached' ? (
              <NewReplyComment
                currFilePath={data.filePath}
                hash={data.hash}
                commentType={this.state.commentType}
              />
            ) : (
              <NewReplyComment
                currFilePath={data.filePath}
                hash={data.hash}
                commentType={this.state.commentType}
                reviewHash={this.props.reviewComment.request.reviewHash}
              />
            ))}
          {data.children && (
            <Button
              size="small"
              endIcon={
                this.state.expandThread ? (
                  <ArrowDropUpIcon />
                ) : (
                  <ArrowDropDownIcon />
                )
              }
              onClick={() => {
                this.setState({ expandThread: !this.state.expandThread });
              }}
              className="threadButton"
            >
              {' '}
              {this.state.expandThread ? 'Hide thread' : 'Show thread'}{' '}
            </Button>
          )}
        </div>
        <div style={style.threadIndent} className="threadList">
          {this.state.expandThread && data.children && (
            <List>
              {data.children.map((reply, index) => {
                if (this.props.detachedComment) {
                  const detached = createDetachedCommentFromJSON(
                    reply,
                    reply.filePath
                  );
                  return (
                    <Comment
                      detachedComment={detached}
                      file={this.props.file}
                      key={index}
                    />
                  );
                } else {
                  const review = createReviewCommentFromJSON(
                    reply,
                    this.props.reviewComment.revision,
                    this.props.reviewComment.request,
                    data.filePath
                  );
                  return (
                    <Comment
                      reviewComment={review}
                      file={this.props.file}
                      key={index}
                    />
                  );
                }
              })}
            </List>
          )}
        </div>
      </>
    );
  }
}
