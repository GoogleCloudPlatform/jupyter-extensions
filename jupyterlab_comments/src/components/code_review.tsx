/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { Divider, Grid, List, Typography } from '@material-ui/core';
import RateReviewIcon from '@material-ui/icons/RateReview';
import { CodeReviewComment, ReviewRequest } from '../service/comment';
import { Comment } from './comment';
import { timeAgo } from '../service/timestamp';
import { RegularFile, NotebookFile } from '../service/file';

interface Props {
  commentsList: Array<CodeReviewComment>;
  reviewRequest: ReviewRequest;
  file: RegularFile | NotebookFile;
}

const style = {
  commentListHeader: {
    fontSize: 19,
    weight: 'bold',
    color: 'black',
    backgroundColor: 'primary',
    paddingLeft: 10,
  },
  commentListDescription: {
    fontSize: 13,
    color: 'grey',
    paddingLeft: 10,
  },
  reviewIcon: {
    paddingLeft: 10,
  },
};

export class CodeReview extends React.Component<Props> {
  render() {
    const comments = this.props.commentsList.map((comment, index) => (
      <div key={index}>
        <Comment reviewComment={comment} file={this.props.file} />
        <Divider />
      </div>
    ));
    const reviewTimeStamp = timeAgo(
      new Date().getTime(),
      this.props.reviewRequest.timestamp
    );
    return (
      <List
        subheader={
          <Grid container direction="column" spacing={0}>
            <Grid container direction="row" spacing={0}>
              <Grid item style={style.reviewIcon}>
                <RateReviewIcon color="primary" />
              </Grid>
              <Grid item>
                <Typography
                  color="primary"
                  style={style.commentListHeader}
                  gutterBottom
                >
                  {'Review requested: ' + this.props.reviewRequest.description}
                </Typography>
              </Grid>
            </Grid>
            <Grid item>
              <Typography
                color="primary"
                style={style.commentListDescription}
                gutterBottom
              >
                {'Requested by: ' + this.props.reviewRequest.requester}
              </Typography>
            </Grid>
            <Grid item>
              <Typography color="primary" style={style.commentListDescription}>
                {'Opened ' + reviewTimeStamp}
              </Typography>
            </Grid>
          </Grid>
        }
      >
        {' '}
        {comments}{' '}
      </List>
    );
  }
}
