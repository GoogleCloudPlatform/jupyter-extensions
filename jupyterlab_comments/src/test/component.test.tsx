import React from 'react';
import { shallow, mount } from 'enzyme';

import { Comment } from '../components/comment';
import { NewReplyComment } from '../components/reply_editor';
// import { ReplyEditor } from '../components/reply_editor';
import { NewCommentThread } from '../components/start_thread';

const getComment = () => {
  const comment = {
    author: 'mkalil',
    text: 'fake comment for testing React component',
    timestamp: '2020',
    range: 'none',
    hash: 'none',
    filePath: 'fake/path',
  };
  return comment;
};

describe('Basic Comment Component Rendering', () => {
  it('detached comment should render correctly', () => {
    const fakeDetachedComment = getComment();
    const componentDetached = shallow(
      <Comment detachedComment={fakeDetachedComment} />
    );

    expect(componentDetached).toMatchSnapshot();
  });

  it('review comment should render correctly', () => {
    const fakeReviewComment = {
      author: 'mkalil',
      text: 'fake comment for testing React component',
      timestamp: '2020',
      range: 'none',
      hash: 'none',
      revision: 'none',
      request: {
        timestamp: '2020',
        reviewRef: 'none',
        targetRef: 'none',
        requester: 'none',
        description: 'none',
        baseCommit: 'none',
        reviewHash: 'none',
      },
      filePath: 'fake/path',
    };
    const componentReview = shallow(
      <Comment reviewComment={fakeReviewComment} />
    );

    expect(componentReview).toMatchSnapshot();
  });
});

describe('Comment Component Expand Threads Button', () => {
  it("has no children comments, don't render thread button", () => {
    const detachedWithoutChildren = getComment();
    const withoutChildren = shallow(
      <Comment detachedComment={detachedWithoutChildren} />
    );
    expect(withoutChildren.find('.thread')).toHaveLength(0);
  });

  it('has children comments, render thread button', () => {
    const detachedWithChildren = {
      author: 'mkalil',
      text: 'fake comment for testing React component',
      timestamp: '2020',
      range: 'none',
      hash: 'none',
      filePath: 'fake/path',
      children: ['fake', 'children', 'list'],
    };
    const withChildren = shallow(
      <Comment detachedComment={detachedWithChildren} />
    );
    const threadButton = withChildren.find('.threadButton');
    expect(threadButton).not.toHaveLength(0);
    expect(threadButton.text().trim()).toEqual('Show thread');
  });

  const detachedWithSingleChild = {
    author: 'mkalil',
    text: 'fake comment for testing React component',
    timestamp: '1543622400',
    range: 'none',
    hash: 'none',
    filePath: 'fake/path',
    children: [
      {
        comment: {
          author: 'test',
          text: 'test',
          timestamp: '1543622400',
          location: {
            range: '0',
          },
        },
        hash: 'rlskjdfo3iruwflkj',
      },
    ],
  };

  const detachedWithTwoChildren = {
    author: 'mkalil',
    text: 'fake comment for testing React component',
    timestamp: '1543622400',
    range: 'none',
    hash: 'none',
    filePath: 'fake/path',
    children: [
      {
        comment: {
          author: 'test',
          text: 'test',
          timestamp: '1543622400',
          location: {
            range: '0',
          },
        },
        hash: 'rlskjdfo3iruwflkj',
      },
      {
        comment: {
          author: 'test',
          text: 'test',
          timestamp: '1543622400',
          location: {
            range: '0',
          },
        },
        hash: 'rlskjdfo3iruwflkj',
      },
    ],
  };
  it('should toggle the comment component state when clicked', () => {
    const withChildren = shallow(
      <Comment detachedComment={detachedWithSingleChild} />
    );
    const threadButton = withChildren.find('.threadButton');
    expect(withChildren.state('expandThread')).toEqual(false);
    threadButton.simulate('click');
    expect(withChildren.state('expandThread')).toEqual(true);
    threadButton.simulate('click');
    expect(withChildren.state('expandThread')).toEqual(false);
  });

  it('should have a single child comment component', () => {
    const withChildren = shallow(
      <Comment detachedComment={detachedWithSingleChild} />
    );
    const threadButton = withChildren.find('.threadButton');
    threadButton.simulate('click');
    const threadList = withChildren.find('.threadList');
    expect(threadList.find('Comment')).toHaveLength(1);
  });

  it('should have two child comment components', () => {
    const withChildren = shallow(
      <Comment detachedComment={detachedWithTwoChildren} />
    );
    const threadButton = withChildren.find('.threadButton');
    threadButton.simulate('click');
    const threadList = withChildren.find('.threadList');
    expect(threadList.find('Comment')).toHaveLength(2);
  });
});

describe('Comment editor components should render correctly', () => {
  it('should render new comment thread editor', () => {
    const newThreadEditor = shallow(
      <NewCommentThread
        serverRoot="fake/path"
        currFilePath="fake/path"
        commentType="detached"
      />
    );
    expect(newThreadEditor).toMatchSnapshot();
  });

  it('should render new reply comment editor', () => {
    const replyEditor = shallow(
      <NewReplyComment
        currFilePath="fake/path"
        hash="hash"
        commentType="detached"
      />
    );
    expect(replyEditor).toMatchSnapshot();
  });
});

describe('Reply button behavior', () => {
  const comment = getComment();
  it('should show comment editor when clicked', () => {
    const component = shallow(<Comment detachedComment={comment} />);
    const replyButton = component.find('.replyButton');
    replyButton.simulate('click');
    const replyEditor = component.find('NewReplyComment');
    expect(replyEditor).toHaveLength(1);
  });

  it('should toggle component state when clicked', () => {
    const component = shallow(<Comment detachedComment={comment} />);
    const replyButton = component.find('.replyButton');
    expect(component.state('showCommentEditor')).toEqual(false);
    replyButton.simulate('click');
    expect(component.state('showCommentEditor')).toEqual(true);
  });
});

describe('Behavior for comment input field', () => {
  it('should display send button for new thread editor', () => {
    const component = mount(
      <NewCommentThread
        serverRoot="root"
        currFilePath="path"
        commentType="review"
      />
    );
    const submit = component.find('.sendThread');
    expect(submit).not.toHaveLength(0);
  });

  it('should display send button for reply editor', () => {
    const component = mount(
      <NewReplyComment hash="hash" currFilePath="path" commentType="detached" />
    );
    const submit = component.find('.sendReply');
    expect(submit).not.toHaveLength(0);
  });

  it('should store the value of the comment (reply)', () => {
    const component = mount(
      <NewReplyComment hash="hash" currFilePath="path" commentType="detached" />
    );
    const input = component.find('.replyCommentTextField').first();
    expect(input.props().value).toEqual('');
    input.props().value = 'new comment';
    expect(input.props().value).toEqual('new comment');
    const submit = component.find('.commentSubmit');
    expect(typeof submit.props().onSubmit === 'function').toBe(true);
  });

  it('should store the value of the comment (new thread)', () => {
    const component = mount(
      <NewCommentThread
        serverRoot="root"
        currFilePath="path"
        commentType="review"
      />
    );
    const input = component.find('.newThreadTextField').first();
    input.props().value = 'new comment';
    expect(input.props().value).toEqual('new comment');
    const submit = component.find('.commentSubmit');
    expect(typeof submit.props().onSubmit === 'function').toBe(true);
  });
});
