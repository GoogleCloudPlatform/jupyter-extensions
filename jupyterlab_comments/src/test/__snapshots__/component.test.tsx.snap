// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Comment Component Rendering detached comment should render correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(ListItem))
    key="context"
  >
    <CommentContext
      filePath="fake/path"
      range="none"
    />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(ListItem))
    alignItems="flex-start"
    key="comment"
  >
    <WithStyles(ForwardRef(ListItemAvatar))>
      <WithStyles(ForwardRef(Avatar))
        alt="avatar"
      />
    </WithStyles(ForwardRef(ListItemAvatar))>
    <WithStyles(ForwardRef(ListItemText))
      primary={
        <div>
          <p
            style={
              Object {
                "display": "inline",
                "fontSize": 15,
                "fontWeight": 500,
                "textAlign": "left",
              }
            }
          >
             
            mkalil
             
          </p>
          <p
            style={
              Object {
                "color": "grey",
                "display": "inline",
                "fontSize": 10,
                "paddingLeft": 10,
                "textAlign": "left",
              }
            }
          >
             
            2020
             
          </p>
        </div>
      }
      secondary={
        <WithStyles(ForwardRef(Typography))
          color="textPrimary"
          style={
            Object {
              "display": "inline",
            }
          }
          variant="body2"
        >
          fake comment for testing React component
        </WithStyles(ForwardRef(Typography))>
      }
    />
  </WithStyles(ForwardRef(ListItem))>
  <div
    style={
      Object {
        "display": "inlineBlock",
        "paddingLeft": 60,
      }
    }
  >
    <WithStyles(ForwardRef(Button))
      className="replyButton"
      color="primary"
      onClick={[Function]}
      size="small"
    >
       
      Reply
       
    </WithStyles(ForwardRef(Button))>
  </div>
  <div
    className="threadList"
    style={
      Object {
        "paddingLeft": 50,
      }
    }
  />
</Fragment>
`;

exports[`Basic Comment Component Rendering review comment should render correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(ListItem))
    key="context"
  >
    <CommentContext
      filePath="fake/path"
      range="none"
    />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(ListItem))
    alignItems="flex-start"
    key="comment"
  >
    <WithStyles(ForwardRef(ListItemAvatar))>
      <WithStyles(ForwardRef(Avatar))
        alt="avatar"
      />
    </WithStyles(ForwardRef(ListItemAvatar))>
    <WithStyles(ForwardRef(ListItemText))
      primary={
        <div>
          <p
            style={
              Object {
                "display": "inline",
                "fontSize": 15,
                "fontWeight": 500,
                "textAlign": "left",
              }
            }
          >
             
            mkalil
             
          </p>
          <p
            style={
              Object {
                "color": "grey",
                "display": "inline",
                "fontSize": 10,
                "paddingLeft": 10,
                "textAlign": "left",
              }
            }
          >
             
            2020
             
          </p>
        </div>
      }
      secondary={
        <WithStyles(ForwardRef(Typography))
          color="textPrimary"
          style={
            Object {
              "display": "inline",
            }
          }
          variant="body2"
        >
          fake comment for testing React component
        </WithStyles(ForwardRef(Typography))>
      }
    />
  </WithStyles(ForwardRef(ListItem))>
  <div
    style={
      Object {
        "display": "inlineBlock",
        "paddingLeft": 60,
      }
    }
  >
    <WithStyles(ForwardRef(Button))
      className="replyButton"
      color="primary"
      onClick={[Function]}
      size="small"
    >
       
      Reply
       
    </WithStyles(ForwardRef(Button))>
  </div>
  <div
    className="threadList"
    style={
      Object {
        "paddingLeft": 50,
      }
    }
  />
</Fragment>
`;

exports[`Comment editor components should render correctly should render new comment thread editor 1`] = `
<CommentEditor
  commentType="detached"
  currFilePath="fake/path"
  serverRoot="fake/path"
/>
`;

exports[`Comment editor components should render correctly should render new reply comment editor 1`] = `
<ReplyEditor
  commentType="detached"
  currFilePath="fake/path"
  hash="hash"
/>
`;
